src/hotspot/cpu/x86/x86_64.ad

```diff
@@ -1472,40 +1472,19 @@ void MachUEPNode::format(PhaseRegAlloc* ra_, outputStream* st) const
 {
   if (UseCompressedClassPointers) {
     st->print_cr("movl    rscratch1, [j_rarg0 + oopDesc::klass_offset_in_bytes()]\t# compressed klass");
-    st->print_cr("\tdecode_klass_not_null rscratch1, rscratch1");
-    st->print_cr("\tcmpq    rax, rscratch1\t # Inline cache check");
+    st->print_cr("\tcmpl    rscratch1, [rax + CompiledICData::speculated_klass_offset()]\t # Inline cache check");
   } else {
-    st->print_cr("\tcmpq    rax, [j_rarg0 + oopDesc::klass_offset_in_bytes()]\t"
-                 "# Inline cache check");
+    st->print_cr("movq    rscratch1, [j_rarg0 + oopDesc::klass_offset_in_bytes()]\t# compressed klass");
+    st->print_cr("\tcmpq    rscratch1, [rax + CompiledICData::speculated_klass_offset()]\t # Inline cache check");
   }
   st->print_cr("\tjne     SharedRuntime::_ic_miss_stub");
-  st->print_cr("\tnop\t# nops to align entry point");
 }
 #endif
 
 void MachUEPNode::emit(CodeBuffer& cbuf, PhaseRegAlloc* ra_) const
 {
   MacroAssembler masm(&cbuf);
-  uint insts_size = cbuf.insts_size();
-  if (UseCompressedClassPointers) {
-    masm.load_klass(rscratch1, j_rarg0, rscratch2);
-    masm.cmpptr(rax, rscratch1);
-  } else {
-    masm.cmpptr(rax, Address(j_rarg0, oopDesc::klass_offset_in_bytes()));
-  }
-
-  masm.jump_cc(Assembler::notEqual, RuntimeAddress(SharedRuntime::get_ic_miss_stub()));
-
-  /* WARNING these NOPs are critical so that verified entry point is properly
-     4 bytes aligned for patching by NativeJump::patch_verified_entry() */
-  int nops_cnt = 4 - ((cbuf.insts_size() - insts_size) & 0x3);
-  if (OptoBreakpoint) {
-    // Leave space for int3
-    nops_cnt -= 1;
-  }
-  nops_cnt &= 0x3; // Do not add nops if code is aligned.
-  if (nops_cnt > 0)
-    masm.nop(nops_cnt);
+  masm.ic_check(CodeEntryAlignment);
```

Comment:
I'm concerned about OptoBreakpoint and friends messing up the alignment:
https://github.com/openjdk/jdk/blob/10beb3184e14e2714cd836029680a8b2d3fd5011/src/hotspot/share/opto/output.cpp#L317

Reply:
It's fine, the breakpoint is insert *after* the prologue.

Reply:
Yeah, that did have me confused as well.